spring-session/src/main/java/org/springframework/session/hazelcast/HazelcastSessionRepository.java

```diff
@@ -0,0 +1,245 @@
+ * </pre>
+ *
+ * In order to support finding sessions by principal name using
+ * {@link #findByIndexNameAndIndexValue(String, String)} method, custom configuration of
+ * {@code IMap} supplied to this implementation is required.
+ *
+ * The following snippet demonstrates how to define required configuration using
+ * programmatic Hazelcast Configuration:
+ *
+ * <pre class="code">
+ * MapAttributeConfig attributeConfig = new MapAttributeConfig()
+ *         .setName(HazelcastSessionRepository.PRINCIPAL_NAME_ATTRIBUTE)
+ *         .setExtractor(PrincipalNameExtractor.class.getName());
+ *
+ * Config config = new Config();
+ *
+ * config.getMapConfig("spring:session:sessions")
+ *         .addMapAttributeConfig(attributeConfig)
+ *         .addMapIndexConfig(new MapIndexConfig(
+ *                 HazelcastSessionRepository.PRINCIPAL_NAME_ATTRIBUTE, false));
+ *
+ * Hazelcast.newHazelcastInstance(config);
+ * </pre>
+ *
+ * This implementation listen for events on the Hazelcast-backed SessionRepository and
+ * translate those events into the corresponding Spring Session events. Publish the Spring
+ * Session events with the given {@link ApplicationEventPublisher}.
+ *
+ * <ul>
+ * <li>entryAdded - {@link SessionCreatedEvent}</li>
+ * <li>entryEvicted - {@link SessionExpiredEvent}</li>
+ * <li>entryRemoved - {@link SessionDeletedEvent}</li>
+ * </ul>
+ *
+ * @author Vedran Pavic
+ * @author Tommy Ludwig
+ * @author Mark Anderson
+ * @since 1.3.0
+ */
+public class HazelcastSessionRepository implements
+		FindByIndexNameSessionRepository<MapSession>,
+		EntryAddedListener<String, MapSession>,
+		EntryEvictedListener<String, MapSession>,
+		EntryRemovedListener<String, MapSession> {
+
+	/**
+	 * The principal name custom attribute name.
+	 */
+	public static final String PRINCIPAL_NAME_ATTRIBUTE = "principalName";
+
+	private static final Log logger = LogFactory.getLog(HazelcastSessionRepository.class);
+
+	private final IMap<String, MapSession> sessions;
+
+	private ApplicationEventPublisher eventPublisher = new ApplicationEventPublisher() {
+
+		public void publishEvent(ApplicationEvent event) {
+		}
+
+		public void publishEvent(Object event) {
+		}
+
+	};
+
+	/**
+	 * If non-null, this value is used to override
+	 * {@link MapSession#setMaxInactiveIntervalInSeconds(int)}.
+	 */
+	private Integer defaultMaxInactiveInterval;
+
+	private String sessionListenerId;
+
+	public HazelcastSessionRepository(IMap<String, MapSession> sessions) {
+		Assert.notNull(sessions, "Sessions IMap must not be null");
+		this.sessions = sessions;
+	}
+
+	@PostConstruct
+	private void init() {
+		this.sessionListenerId = this.sessions.addEntryListener(this, true);
+	}
+
+	@PreDestroy
+	private void close() {
+		this.sessions.removeEntryListener(this.sessionListenerId);
+	}
+
+	/**
+	 * Sets the {@link ApplicationEventPublisher} that is used to publish
+	 * {@link SessionDestroyedEvent}. The default is to not publish a
```

Comment:
This is the publisher for all of the Session events, but the JavaDoc here only mentions the `SessionDestroyedEvent`, which I think might be a little confusing for users.